pkg/generators/openapi_json.go

```diff
@@ -0,0 +1,292 @@
+		if err != nil {
+			return nil, err
+		}
+	case types.Struct, types.Interface:
+		props, req, err := g.generateMembers(t, false)
+		if err != nil {
+			return nil, err
+		}
+		schema.Properties = props
+		schema.SchemaProps.Required = req
+	default:
+		return nil, fmt.Errorf("cannot generate spec for type %v", t)
+	}
+	schema.SchemaProps.Description = strings.TrimSpace(parseDescription(m.CommentLines))
+	schema.VendorExtensible = *extensions
+	return schema, nil
+}
+
+func (g openAPIGeneratorJSON) generateMapProperty(t *types.Type) (*spec.Schema, error) {
+	var err error
+	keyType := resolveAliasAndPtrType(t.Key)
+	elemType := resolveAliasAndPtrType(t.Elem)
+
+	// According to OpenAPI examples, only map from string is supported
+	if keyType.Name.Name != "string" {
+		return nil, fmt.Errorf("map with non-string keys are not supported by OpenAPI in %v", t)
+	}
+	typeString, format := openapi.GetOpenAPITypeFormat(elemType.String())
+	if typeString != "" {
+		schema := &spec.Schema{}
+		schema.Type = []string{"object"}
+		schema.AdditionalProperties = &spec.SchemaOrBool{
+			Schema: g.generateSimpleProperty(typeString, format),
+		}
+		return schema, nil
+	}
+
+	// we don't set any fields on this variable until after the switch so that
+	// we can easily call `schema, err = g.generateXYZ()` and set the fields we
+	// need at the end
+	schema := &spec.Schema{}
+	switch elemType.Kind {
+	case types.Builtin:
+		return nil, fmt.Errorf("please add type %v to getOpenAPITypeFormat function", elemType)
+	case types.Struct:
+		props, req, err := g.generateMembers(elemType, false)
+		if err != nil {
+			return nil, err
+		}
+		schema.AdditionalProperties = &spec.SchemaOrBool{
+			Schema: &spec.Schema{
+				SchemaProps: spec.SchemaProps{
+					Properties: props,
+					Required:   req,
+				},
+			},
+		}
+	case types.Slice, types.Array:
+		schema, err = g.generateSliceProperty(elemType)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("map Element kind %v is not supported in %v", elemType.Kind, t.Name)
+	}
+	schema.Type = []string{"object"}
+	return schema, nil
+}
+
+func (g openAPIGeneratorJSON) generateSliceProperty(t *types.Type) (*spec.Schema, error) {
+	elemType := resolveAliasAndPtrType(t.Elem)
+	typeString, format := openapi.GetOpenAPITypeFormat(elemType.String())
+	if typeString != "" {
+		schema := &spec.Schema{}
+		schema.Type = []string{"array"}
+		schema.AdditionalProperties = &spec.SchemaOrBool{
+			Schema: g.generateSimpleProperty(typeString, format),
+		}
+		return schema, nil
+	}
+	schema := &spec.Schema{}
+	switch elemType.Kind {
+	case types.Builtin:
+		return nil, fmt.Errorf("please add type %v to getOpenAPITypeFormat function", elemType)
+	case types.Struct:
+		props, req, err := g.generateMembers(elemType, false)
+		if err != nil {
+			return nil, err
+		}
+		glog.V(1).Infof("Got member props: %+v", props)
+		schema.AdditionalProperties = &spec.SchemaOrBool{Schema: &spec.Schema{}}
```

Comment:
why do we create AdditionalProperties for members?

Reply:
for slices there is `items`, isn't it? I.e. `items` is for slices what `additionalPropoerties` is for structs.

Reply:
[Reference](https://arxiv.org/pdf/1701.02221.pdf)
Added if it makes sense. @sttts 
**Array schemas** 
 Array schemas are specified with the "type":  "array" keyword. For arrays there are two ways of specifying what kind of documents we find in arrays. We can use a pair
"items" : [J1,... JN] to specify a document with an array of n elements, where each i-th element must satisfy schema Ji.**We can also use "additionalItems":J to specify that all elements in the array must satisfy schema J.** If both keywords are used together, then we allow the array to have more values than those specified with items, as long as they agree with the schema specified in additionalItems.
For example, the following schema validates against arrays of at least 2 elements, where the first two are strings and the remaining ones, if they exists, are numbers.
{
"type": "array",
"items": [{"type":"string"}, {"type":"string"}],
"additionalItems": {"type":"number"},
"uniqueItems":true
}